hotfix for memory leak in aot autograd induced by saving tensors for backward

bdhirsh opened the pull request:
Workaround fix in AOTAutograd for https://github.com/pytorch/pytorch/issues/94990 (see the comments for more details / discussion)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95392
* #95421
* __->__ #95101

bdhirsh committed:
hotfix for memory leak in aot autograd induced by saving tensors for backward

[ghstack-poisoned]

bdhirsh committed:
Update on "hotfix for memory leak in aot autograd induced by saving tensors for backward"

Workaround fix in AOTAutograd for https://github.com/pytorch/pytorch/issues/94990 (see the comments for more details / discussion)




[ghstack-poisoned]

Chillee reviewed (commented):
I'm worried this will have performance ramifications - we're running a detach call per activation saved.

albanD commented:
> I'm worried this will have performance ramifications - we're running a detach call per activation saved.

We do a lot of detach internally haha, I really don't think this will change anything unless some FakeTensorMode weirdness kicks in,.

albanD reviewed (approved):
SGTM

bdhirsh commented:
I'll try to get some benchmarks just to be safe

bdhirsh commented:
Some sanity benchmarks:

`x.detach()` cost: ~7us (per tensor)

cost of running just the forward of a no-op AOTAutograd function: ~110us

cost of running the forward + backward of a no-op AOTAutograd function: ~540us

no-op code:
```
from functorch.compile import aot_function, nop

def f(x):
    return x

a = torch.ones(1, device='cuda', requires_grad=True)
f_ = aot_function(f, fw_compiler=nop)

t0 = benchmark.Timer(
    stmt='out = f_(a); out.sum().backward()',
    globals={'f_': f_, 'a': a}
)
print(t0.timeit(100))
```

Alban also convinced me that trying to fix this ref cycle directly in autograd (aka option b that I put in the issue) is pretty non-trivial / might require doing some extra calculations in the general case eager hot-path :(. LMK if anyone has strong opinions. That leads me towards taking on the minor overhead for now for the release, and ideally fixing this ref cycle using option (a) in the longer term.

I also need to make some other fixes since this PR broke tests - I need to properly fix up the requires_grad field of the saved tensors in the backward, so we don't lose the nice error message that you're supposed to get when you try to double backward.

Chillee commented:
> cost of running just the forward of a no-op AOTAutograd function: ~110us

Jeez. That seems pretty bad lol. I think this has gotten worse 🤔

bdhirsh committed:
Update on "hotfix for memory leak in aot autograd induced by saving tensors for backward"

Workaround fix in AOTAutograd for https://github.com/pytorch/pytorch/issues/94990 (see the comments for more details / discussion)




[ghstack-poisoned]

gchanan commented on test/functorch/test_aotdispatch.py:
```diff
@@ -1016,6 +1016,34 @@ def inp_callable(req_grad):
+        # But AOTAutograd makes it more likely to show up from tracing user programs,
+        # so we deal with it by manually detaching the tensors that we save for backward.
+        # This is completely wrong and would give wrong results if we were to do double backward.
+        # Fortunately today, double backward is explicitly banned in AOTAutograd.
```
would we obviously trigger this issue if we tried to run double backwards?

bdhirsh replied:
just checking- you're asking if whether or not we will fail loudly with a nice error message if you try to run double backwards today (and not e.g. silently get wrong results / get a bad error message)?

Jeffrey did some work a few weeks ago to make sure we get a nice error message. I actually just updated this PR to make sure that we still get the nice error (there's an expect test that asserts we get a good error message today).

bdhirsh replied:
Oh, you're wondering if this ref cycle can show up more commonly if you're doing double backward with an `autograd.Function`?

I'm not 100% sure, but I'm leaning towards no - only this specific usage of autograd.Function will cause the mem leak (regardless of whether you're doing single or double backward).

bdhirsh committed:
Update on "hotfix for memory leak in aot autograd induced by saving tensors for backward"

Workaround fix in AOTAutograd for https://github.com/pytorch/pytorch/issues/94990 (see the comments for more details / discussion)




[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2180,7 +2180,8 @@ def forward(ctx, *deduped_flat_tensor_args):
                 )
-                ctx.save_for_backward(*tensors_saved_for_backwards)
+                # See Note [Detaching saved tensors in AOTAutograd]
+                ctx.save_for_backward(*map(lambda x: x.detach() if x._is_view() else x, tensors_saved_for_backwards))
```
@Chillee this should now be a bit less egregious, and still fixes the mem leak 🙂 (we only need to worry about the case when our saved-for-bw tensor is a differentiable view)

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "hotfix for memory leak in aot autograd induced by saving tensors for backward"

Workaround fix in AOTAutograd for https://github.com/pytorch/pytorch/issues/94990 (see the comments for more details / discussion)




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
